crates/accelerate/src/basis/basis_translator/compose_transforms.rs

```diff
@@ -0,0 +1,206 @@
+            gate_obj.operation,
+            &qubits,
+            &[],
+            if gate_obj.params.is_empty() {
+                None
+            } else {
+                Some(gate_obj.params)
+            },
+            gate_obj.extra_attrs,
+            #[cfg(feature = "cache_pygates")]
+            Some(gate.into()),
+        )?;
+        mapped_instructions.insert(
+            (gate_name.clone(), gate_num_qubits),
+            (placeholder_params, dag),
+        );
+
+        for (gate_name, gate_num_qubits, equiv_params, equiv) in basis_transforms {
+            for ((_mapped_instr_name, _), (_dag_params, dag)) in &mut mapped_instructions {
+                let doomed_nodes = dag
+                    .op_nodes(true)
+                    .filter_map(|node| {
+                        if let Some(NodeType::Operation(op)) = dag.dag().node_weight(node) {
+                            if (gate_name.as_str(), *gate_num_qubits)
+                                == (op.op.name(), op.op.num_qubits())
+                            {
+                                Some((
+                                    node,
+                                    op.params_view()
+                                        .iter()
+                                        .map(|x| x.clone_ref(py))
+                                        .collect::<SmallVec<[Param; 3]>>(),
+                                ))
+                            } else {
+                                None
+                            }
+                        } else {
+                            None
+                        }
+                    })
+                    .collect::<Vec<_>>();
+                for (node, params) in doomed_nodes {
+                    let param_mapping: HashMap<ParameterUuid, Param> = equiv_params
+                        .iter()
+                        .map(|x| ParameterUuid::from_parameter(x.to_object(py).bind(py)))
+                        .zip(params)
+                        .map(|(uuid, param)| -> PyResult<(ParameterUuid, Param)> {
+                            Ok((uuid?, param.clone_ref(py)))
+                        })
+                        .collect::<PyResult<_>>()?;
+                    let mut replacement = equiv.clone();
+                    replacement
+                        .0
+                        .assign_parameters_from_mapping(py, param_mapping)?;
+                    let replace_dag: DAGCircuit =
+                        DAGCircuit::from_circuit_data(py, replacement.0, true)?;
+                    let op_node = dag.get_node(py, node)?;
+                    dag.substitute_node_with_dag(py, op_node.bind(py), &replace_dag, None, true)?;
+                }
+            }
+        }
+    }
+    Ok(mapped_instructions)
+}
+
+fn get_example_gates(
+    py: Python,
+    dag: &DAGCircuit,
+    example_gates: Option<Box<HashMap<(String, u32), PackedInstruction>>>,
+) -> PyResult<Box<HashMap<(String, u32), PackedInstruction>>> {
```

Comment:
The original Python function is _strange_. A more appropriate way to do this in Rust would be to require that the caller allocates the map and provides a mutable reference to it so all recursive calls to the function can keep appending to it.

```suggestion
    example_gates: &mut HashMap<(String, u32), PackedInstruction>,
) -> PyResult<()> {
```